In [1]:
import keras
from keras import callbacks
from keras.datasets import mnist
from keras.models import *
from keras.layers import *
from keras.optimizers import *
from keras import backend as K
from sklearn import preprocessing
import os

import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
plt.style.use('ggplot')

Using TensorFlow backend.


In [2]:
os.getcwd()

'/home/nvaitc/Documents/repos/ML-RN-Project'

## 1. Generate x_test and y_test

In [3]:
RN_input_test = np.load('RN_input_test.npy')
a_test = np.load('a_test.npy')
print (RN_input_test.shape)
print (a_test.shape)

(995, 190, 1576)
(995,)


In [4]:
le_test = preprocessing.LabelEncoder()
le_test.fit(a_test)
test_labels = le_test.transform(a_test)

l = len(np.unique(test_labels))
print(l)

one_hot_labels = keras.utils.to_categorical(test_labels, num_classes=l)
print(one_hot_labels)

6
[[0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 ...
 [0. 0. 0. 0. 0. 1.]
 [0. 0. 0. 0. 0. 1.]
 [0. 1. 0. 0. 0. 0.]]


In [5]:
x_test = RN_input_test
y_test = one_hot_labels

print(x_test.shape)
print(y_test.shape)

(995, 190, 1576)
(995, 6)


# Note:

- All models are run with batch size = 64, epoch = 200, validation split = 10% (unless otherwise stated).

- Original model: g = [256,256,256,256] and f = [256,512,6] with 50% dropout rate at 2nd layer of f. ReLu activation for all except last layer of f, which is softmax. Before passing the output from g to f, it is summed over 190 combinations.

## 2. Comparison between different sample sizes

(make a table to compare training and validation accuracy)

...

Embedded c without labels: Validation Accuracy = ~74-75%

Embedded c with labels: Validation Accuracy = ~85-85%

Conclusion: It is important to include labels.

## 3. Comparison between different RN inputs

(make a table to compare training and validation accuracy)

...

## 3. Comparison between different modification of models [with RN inputs (original)]

(a) Original model ( g = [256 x 4], f = [256,512,6] )

(b) Increase number of layers in g from 4 to 6 ( g = [256 x 6], f = [256,512,6] )

(c) Increase number of layers in g from 4 to 8 ( g = [256 x 8], f = [256,512,6] )

(d) Increase number of neurons in each layer of g from 256 to 512 ( g = [512 x 4], f = [256,512,6] )

(e) Increase number of neurons in each layer of g from 256 to 512 and 1st layer of f from 256 to 512 ( g = [512 x 4], f = [512,512,6] )

(f) Increase number of neurons in each layer of g from 256 to 512. And increase number of layers in g from 4 to 6. ( g = [512 x 6], f = [256,512,6] )

(g) Increase number of neurons in each layer of g from 256 to 512 and 1st layer of f from 256 to 512. And increase number of layers in g from 4 to 6. ( g = [512 x 6], f = [512,512,6] )

(h) Double the number of neurons in all layers except for the last layer of f. Increase the number of layers of g from 4 to 6. ( g = [512 x 6], f = [512,1024,6] )

### (a) Original model ( g = [256,256,256,256], f = [256,512,6] )

In [7]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_ori')

In [8]:
#load json and create model_ori

model_ori_json_file = open("model_ori.json","r")
loaded_model_ori_json = model_ori_json_file.read()
model_ori_json_file.close()
model_ori = model_from_json(loaded_model_ori_json)

In [9]:
#load weights into new model_ori

model_ori.load_weights("W_ori_159_0.87.h5")
print("Loaded model from disk")

Loaded model from disk


In [10]:
#evaluate loaded model on test data

model_ori.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_ori = model_ori.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_ori: 100%")
print("Validation Accuracy on model_ori: 87.1%")
print("Test Accuracy on model_ori: ", "{0:0.1f}".format(score_ori[1]*100), "%")

Training Accuracy on model_ori: 100%
Validation Accuracy on model_ori: 87.1%
Test Accuracy on model_ori:  85.5 %


In [11]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (b) Increase number of layers in g from 4 to 6 ( g = [256,256,256,256,256,256], f = [256,512,6] )

In [6]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_g6')

In [7]:
#load json and create model_ori

model_g6_json_file = open("model_g6.json","r")
loaded_model_g6_json = model_g6_json_file.read()
model_g6_json_file.close()
model_g6 = model_from_json(loaded_model_g6_json)

In [8]:
#load weights into new model_ori

model_g6.load_weights("W_g6_188_0.89.h5")
print("Loaded model from disk")

Loaded model from disk


In [10]:
#evaluate loaded model on test data

model_g6.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_g6 = model_g6.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_g6: 100%")
print("Validation Accuracy on model_g6: 89.1%")
print("Test Accuracy on model_g6: ", "{0:0.1f}".format(score_g6[1]*100), "%")

Training Accuracy on model_g6: 100%
Validation Accuracy on model_g6: 89.1%
Test Accuracy on model_g6:  88.4 %


In [11]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (c) Increase number of layers in g from 4 to 8 ( g = [256 x 8], f = [256,512,6] )

In [12]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_g8')

In [13]:
#load json and create model_ori

model_g8_json_file = open("model_g8.json","r")
loaded_model_g8_json = model_g8_json_file.read()
model_g8_json_file.close()
model_g8 = model_from_json(loaded_model_g8_json)

In [14]:
#load weights into new model_ori

model_g8.load_weights("W_g8_05_0.22.h5")
print("Loaded model from disk")

Loaded model from disk


In [15]:
#evaluate loaded model on test data

model_g8.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_g8 = model_g8.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_g8: 18.4%")
print("Validation Accuracy on model_g8: 22.2%")
print("Test Accuracy on model_g8: ", "{0:0.1f}".format(score_g8[1]*100), "%")

Training Accuracy on model_g8: 18.4%
Validation Accuracy on model_g8: 22.2%
Test Accuracy on model_g8:  20.0 %


In [16]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (d) Increase number of neurons in each layer of g from 256 to 512 ( g = [512 x 4], f = [256,512,6] )

In [17]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_512')

In [18]:
#load json and create model_ori

model_512_json_file = open("model_512.json","r")
loaded_model_512_json = model_512_json_file.read()
model_512_json_file.close()
model_512 = model_from_json(loaded_model_512_json)

In [19]:
#load weights into new model_ori

model_512.load_weights("W_512_179_0.88.h5")
print("Loaded model from disk")

Loaded model from disk


In [20]:
#evaluate loaded model on test data

model_512.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_512 = model_512.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_512: 100%")
print("Validation Accuracy on model_512: 88.2%")
print("Test Accuracy on model_512: ", "{0:0.1f}".format(score_512[1]*100), "%")

Training Accuracy on model_512: 100%
Validation Accuracy on model_512: 88.2%
Test Accuracy on model_512:  85.0 %


In [21]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (e) Increase number of neurons in each layer of g from 256 to 512 and 1st layer of from 256 to 512 ( g = [512 x 4], f = [512,512,6] )

In [22]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_512f512')

In [23]:
#load json and create model_ori

model_512f512_json_file = open("model_512f512.json","r")
loaded_model_512f512_json = model_512f512_json_file.read()
model_512f512_json_file.close()
model_512f512 = model_from_json(loaded_model_512f512_json)

In [24]:
#load weights into new model_ori

model_512f512.load_weights("W_512f512_193_0.88.h5")
print("Loaded model from disk")

Loaded model from disk


In [25]:
#evaluate loaded model on test data

model_512f512.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_512f512 = model_512f512.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_512f512: 99.9%")
print("Validation Accuracy on model_512f512: 87.9%")
print("Test Accuracy on model_512f512: ", "{0:0.1f}".format(score_512f512[1]*100), "%")

Training Accuracy on model_512f512: 99.9%
Validation Accuracy on model_512f512: 87.9%
Test Accuracy on model_512f512:  88.3 %


In [26]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (f) Increase number of neurons in each layer of g from 256 to 512. And increase number of layers in g from 4 to 6. ( g = [512 x 6], f = [256,512,6] )

In [27]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_g6x512')

In [28]:
#load json and create model_ori

model_g6x512_json_file = open("model_g6x512.json","r")
loaded_model_g6x512_json = model_g6x512_json_file.read()
model_g6x512_json_file.close()
model_g6x512 = model_from_json(loaded_model_g6x512_json)

In [29]:
#load weights into new model_ori

model_g6x512.load_weights("W_g6x512_164_0.88.h5")
print("Loaded model from disk")

Loaded model from disk


In [30]:
#evaluate loaded model on test data

model_g6x512.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_g6x512 = model_g6x512.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_g6x512: 100%")
print("Validation Accuracy on model_g6x512: 87.8%")
print("Test Accuracy on model_g6x512: ", "{0:0.1f}".format(score_g6x512[1]*100), "%")

Training Accuracy on model_g6x512: 100%
Validation Accuracy on model_g6x512: 87.8%
Test Accuracy on model_g6x512:  88.5 %


In [31]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (g) Increase number of neurons in each layer of g from 256 to 512 and 1st layer of f from 256 to 512. And increase number of layers in g from 4 to 6. ( g = [512 x 6], f = [512,512,6] )

In [32]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_g6x512f512')

In [33]:
#load json and create model_ori

model_g6x512f512_json_file = open("model_g6x512f512.json","r")
loaded_model_g6x512f512_json = model_g6x512f512_json_file.read()
model_g6x512f512_json_file.close()
model_g6x512f512 = model_from_json(loaded_model_g6x512f512_json)

In [34]:
#load weights into new model_ori

model_g6x512f512.load_weights("W_g6x512f512_179_0.88.h5")
print("Loaded model from disk")

Loaded model from disk


In [35]:
#evaluate loaded model on test data

model_g6x512f512.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_g6x512f512 = model_g6x512f512.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_g6x512f512: 99.7%")
print("Validation Accuracy on model_g6x512f512: 88.3%")
print("Test Accuracy on model_g6x512f512: ", "{0:0.1f}".format(score_g6x512f512[1]*100), "%")

Training Accuracy on model_g6x512f512: 99.7%
Validation Accuracy on model_g6x512f512: 88.3%
Test Accuracy on model_g6x512f512:  89.0 %


In [36]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project')

### (h) Double the number of neurons in all layers except for the last layer of f. Increase the number of layers of g from 4 to 6. ( g = [512 x 6], f = [512,1024,6] )

In [37]:
os.chdir('/home/nvaitc/Documents/repos/ML-RN-Project/model_double')

In [38]:
#load json and create model_ori

model_double_json_file = open("model_double.json","r")
loaded_model_double_json = model_double_json_file.read()
model_double_json_file.close()
model_double = model_from_json(loaded_model_double_json)

In [39]:
#load weights into new model_ori

model_double.load_weights("W_double_177_0.88.h5")
print("Loaded model from disk")

Loaded model from disk


In [40]:
#evaluate loaded model on test data

model_double.compile(optimizer = Adam(lr = 2e-4), loss = 'categorical_crossentropy', metrics = ['accuracy'])
score_double = model_double.evaluate(x_test,y_test,verbose = 0)
print("Training Accuracy on model_double: 100%")
print("Validation Accuracy on model_double: 88.2%")
print("Test Accuracy on model_double: ", "{0:0.1f}".format(score_double[1]*100), "%")

Training Accuracy on model_double: 100%
Validation Accuracy on model_double: 88.2%
Test Accuracy on model_double:  88.4 %


# Experiments and Results



We have made several comparison between different types of embeddings, different sample sizes, different types of inputs to feed to the RN module and different modifications to the RN module with our implementation of USE embeddings. We compared the accuracy of the test set for different modifications to the RN module and compared the accuracy of the validation set for the rest of the comparisons.

For the experiments, all models were run with batch size of 64, with 200 epochs and validation set was taken from the last 10\% of the training set. For the original RN module, $f_\phi \left(\sum_{i,j} g_\theta (o_i,o_j,q)\right)$, $g_\theta$ is a four-layer MLP 256 unites per layer and $f_\phi$ is a three-layer MLP consisting of 256, 512 and 6 units, where the final softmax output was optimized with a cross-entropy loss function using the Adam optimizer with a learning rate of $2e^{-4}$. All models are run on original RN module except for the different modifications to the RN module.



## Comparison between Different Embeddings



We have made comparisons between the two different embeddings of the sentences of the contexts, questions and answers and below are the result:

$

\begin{tabular}{|c|c|c|}

\hline
&&\\
\hline
&&\\
\hline

\end{tabular}

$